pkg/client/reader_test.go

```diff
@@ -0,0 +1,230 @@
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/proto"
+)
+
+func TestNewFromReader_Empty(t *testing.T) {
+	_, err := NewFromReader(bytes.NewBuffer([]byte("")), nil)
+	assert.Error(t, err)
+}
+
+func TestNewFromReader_BadProto(t *testing.T) {
+	_, err := NewFromReader(bytes.NewBuffer([]byte("bad proto")), nil)
+	assert.Error(t, err)
+}
+
+func TestNewFromReader_Connects(t *testing.T) {
+	ca, err := NewCA()
+	require.NoError(t, err)
+	peer, err := ca.GeneratePair()
+	require.NoError(t, err)
+	peerCert, err := peer.Certificate()
+	require.NoError(t, err)
+
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(ca.caPEM)
+	creds := credentials.NewTLS(&tls.Config{
+		Certificates: []tls.Certificate{peerCert},
+		ClientCAs:    caCertPool,
+		ClientAuth:   tls.RequireAndVerifyClientCert,
+	})
+
+	var m sync.Mutex
+	token := "expected_token"
+	gotValid := false
+	srv := StubServer{
+		CheckinImpl: func(observed *proto.StateObserved) *proto.StateExpected {
+			m.Lock()
+			defer m.Unlock()
+
+			if observed.Token == token {
+				gotValid = true
+				return &proto.StateExpected{
+					State:          proto.StateExpected_RUNNING,
+					ConfigStateIdx: 1,
+					Config:         "config",
+				}
+			}
+			// disconnect
+			return nil
+		},
+		ActionImpl: func(response *proto.ActionResponse) error {
+			// actions not tested here
+			return nil
+		},
+		actions: make(chan *performAction, 100),
+	}
+	require.NoError(t, srv.Start(grpc.Creds(creds)))
+	defer srv.Stop()
+
+	connInfo := &proto.ConnInfo{
+		Addr:     fmt.Sprintf(":%d", srv.Port),
+		Token:    token,
+		CaCert:   ca.caPEM,
+		PeerCert: peer.Crt,
+		PeerKey:  peer.Key,
+	}
+	connInfoBytes, err := protobuf.Marshal(connInfo)
+	require.NoError(t, err)
+	connInfoReader := bytes.NewReader(connInfoBytes)
+
+	impl := &StubClientImpl{}
+	validClient, err := NewFromReader(connInfoReader, impl)
+	require.NoError(t, err)
+	require.NoError(t, validClient.Start(context.Background()))
+	defer validClient.Stop()
+	require.NoError(t, waitFor(func() error {
+		m.Lock()
+		defer m.Unlock()
+
+		if !gotValid {
+			return fmt.Errorf("server never recieved valid token")
+		}
+		return nil
+	}))
+}
+
+type CertificateAuthority struct {
+	caCert     *x509.Certificate
+	privateKey crypto.PrivateKey
+	caPEM      []byte
+}
+
+type Pair struct {
+	Crt []byte
+	Key []byte
+}
+
+func NewCA() (*CertificateAuthority, error) {
+	ca := &x509.Certificate{
+		SerialNumber: big.NewInt(1653),
+		Subject: pkix.Name{
+			Organization: []string{"elastic-fleet"},
+			CommonName:   "localhost",
+		},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	privateKey, _ := rsa.GenerateKey(rand.Reader, 2048)
```

Comment:
I don't see a good reason an error should happen but probably still worth checking for it? If one happens, things for sure explode further down I guess.

Reply:
Yeah it would explode because privateKey would be empty.